<a href="https://colab.research.google.com/github/PKGISREAL/Big_data/blob/main/Homework_pandas_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Выполнил: Округин Никита Анатольевич


<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#-Homework-1-(Topics-1-4:-Data-Structures,-Indexing,-Cleaning-Data,-Math)" data-toc-modified-id="-Homework-1-(Topics-1-4:-Data-Structures,-Indexing,-Cleaning-Data,-Math)-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)</center></a></span><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Task-1." data-toc-modified-id="Task-1.-1.0.0.1"><span class="toc-item-num">1.0.0.1&nbsp;&nbsp;</span>Task 1.</a></span></li><li><span><a href="#Варианты-ответов:" data-toc-modified-id="Варианты-ответов:-1.0.0.2"><span class="toc-item-num">1.0.0.2&nbsp;&nbsp;</span>Варианты ответов:</a></span></li><li><span><a href="#Task-2." data-toc-modified-id="Task-2.-1.0.0.3"><span class="toc-item-num">1.0.0.3&nbsp;&nbsp;</span>Task 2.</a></span></li><li><span><a href="#Вопросы:" data-toc-modified-id="Вопросы:-1.0.0.4"><span class="toc-item-num">1.0.0.4&nbsp;&nbsp;</span>Вопросы:</a></span></li><li><span><a href="#Task-3." data-toc-modified-id="Task-3.-1.0.0.5"><span class="toc-item-num">1.0.0.5&nbsp;&nbsp;</span>Task 3.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.0.0.6"><span class="toc-item-num">1.0.0.6&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-4." data-toc-modified-id="Task-4.-1.0.0.7"><span class="toc-item-num">1.0.0.7&nbsp;&nbsp;</span>Task 4.</a></span></li><li><span><a href="#Пояснения:" data-toc-modified-id="Пояснения:-1.0.0.8"><span class="toc-item-num">1.0.0.8&nbsp;&nbsp;</span>Пояснения:</a></span></li></ul></li><li><span><a href="#Информация-для-последующих-заданий" data-toc-modified-id="Информация-для-последующих-заданий-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>Информация для последующих заданий</a></span></li></ul></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Таблица-transactions.csv" data-toc-modified-id="Таблица-transactions.csv-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Таблица <code>transactions.csv</code></a></span></li><li><span><a href="#Таблица-gender_train.csv" data-toc-modified-id="Таблица-gender_train.csv-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Таблица <code>gender_train.csv</code></a></span></li><li><span><a href="#Таблица-tr_mcc_codes.csv" data-toc-modified-id="Таблица-tr_mcc_codes.csv-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Таблица <code>tr_mcc_codes.csv</code></a></span></li><li><span><a href="#Таблица-tr_types.csv" data-toc-modified-id="Таблица-tr_types.csv-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Таблица <code>tr_types.csv</code></a></span><ul class="toc-item"><li><span><a href="#Task-5." data-toc-modified-id="Task-5.-1.1.4.1"><span class="toc-item-num">1.1.4.1&nbsp;&nbsp;</span>Task 5.</a></span></li><li><span><a href="#Task-6." data-toc-modified-id="Task-6.-1.1.4.2"><span class="toc-item-num">1.1.4.2&nbsp;&nbsp;</span>Task 6.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.1.4.3"><span class="toc-item-num">1.1.4.3&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-7." data-toc-modified-id="Task-7.-1.1.4.4"><span class="toc-item-num">1.1.4.4&nbsp;&nbsp;</span>Task 7.</a></span></li><li><span><a href="#Выберите-все-верные-пункты:" data-toc-modified-id="Выберите-все-верные-пункты:-1.1.4.5"><span class="toc-item-num">1.1.4.5&nbsp;&nbsp;</span>Выберите все верные пункты:</a></span></li><li><span><a href="#Task-8." data-toc-modified-id="Task-8.-1.1.4.6"><span class="toc-item-num">1.1.4.6&nbsp;&nbsp;</span>Task 8.</a></span></li></ul></li></ul></li></ul></li></ul></div>

# <center> Homework 1 (Topics 1-4: Data Structures, Indexing, Cleaning Data, Math)

In [1]:
import pandas as pd
import numpy as np

#### Task 1.
Задание:

1. Создайте Series из последовательности 15 значений, равномерно разбивающих отрезок [0, 20] (воспользуйтесь функцией linspace)
2. Определите отношение элементов полученной серии к их предыдущим элементам (\*).
3. В результате необходимо получить среднее полученного вектора, оставив в нём только те значения, которые не более чем 1.5 (\**).

Выберите из ответов тот, который максимально близок к полученному (c точки зрения абсолютной разницы).

#### Варианты ответов:
    - 1) 1.24
    - 2) 1.18
    - 3) 0.71
    - 4) 1.13


##### Пояснения:
(\*) Если было бы необходимо найти последовательность из 3-х значений,  
равномерно разбивающих отрезок [0,1], то это были бы значения [0, 0.5, 1].

(\**) Если был бы дан список элементов a = [1,2,3,12], отношения элементов к предыдущим будут равны [NaN, 2, 1.5, 4].  
А на последнем этапе в таком примере останется только [1.5] и среднее значение будет также 1.5.

In [2]:
### Type your code here
series = pd.Series(np.linspace(0, 20, 15))
ratio_series = series/series.shift(1)
print(ratio_series[ratio_series <= 1.5].mean())



1.181677812927813


**Ответ: Б**

#### Task 2.
Выберите все верные ответы касательно следующих 3-х Series:
    - pd.Series('abcde'); (1)
    - pd.Series(['abcde']); (2)
    - pd.Series(list('abcde')); (3)
    - pd.Series("abcde"); (4)
    
##### Пояснения:
(\*)
- функция list: в строке каждый символ - это отдельный элемент для list
- квадратные скобки: в квадратных скобках списку передается множество элементов по отдельности через запятую


#### Вопросы:
    - 1) Серия (1) совпадает с серией (2), так как в каждом из случаев серия создаётся из списка строк
    - 2) Серия (2) совпадает с серией (3), так как в каждом из случаев серия создаётся из списка символов
    - 3) Серия (1) не совпадает с серией (4), так как в (4) используются двойные кавычки "" вместо одинарных ''

Рассмотрим со стороны кода


In [3]:

series_1 = pd.Series('abcde')
series_2 = pd.Series(['abcde'])
series_3 = pd.Series(list('abcde'))
series_4 = pd.Series("abcde")

print("Series №1")
print(series_1, end = '\n \n')

print("Series №2")
print(series_2, end = '\n \n')

print("Series №3")
print(series_3, end = '\n \n')

print("Series №4")
print(series_4, end = '\n \n')

Series №1
0    abcde
dtype: object
 
Series №2
0    abcde
dtype: object
 
Series №3
0    a
1    b
2    c
3    d
4    e
dtype: object
 
Series №4
0    abcde
dtype: object
 


**Исходя из результатов, правильными ответами считаются 1 и 3**

#### Task 3.

По клиенту получены зашумленные данные (объект s типа Series) по его транзакциям.

Для заданного ниже объекта s проделайте следующее:
1. Создайте новый Series, значения которого совпадают со значениями s, а индексы - целочисленные значения от 2 до 12, не включая 12.
2. Выберите из s элементы с индексами 3 и 5, после чего просуммируйте их, сохранив результат (1).
3. Выберите из s только целочисленные элементы и вычислите их дисперсию (2). (\*)

Все полученные результаты округлите до 2-х знаков после запятой.

#### Выберите все верные пункты:
    - 1) Ответ (1) - 642.52
    - 2) Ответ (1) - 91.78
    - 3) Ответ (1) - не может быть определён
    - 4) Ответ (2) - 57591.19
    - 5) Ответ (2) - 210.12
    - 6) Ответ (2) - не может быть определён

##### Пояснения:
(\*) Целочисленные значения - значения, имеющие тип int.
- Дисперсия рассчитывается с помощью функции из библиотеки numpy: np.var( , ddof=0) или встроенной в python функции: .var(ddof=1)

In [4]:
s = pd.Series(data=['1', 2, 3.1, 'hi!', 5, -512, 12.42, 'sber', 10.10, 98],
              index=range(6, 26, 2))
### Type your code here
new_s = pd.Series(s.values, index=range(2, 12))

# ОШИБКА
#sum_elements = s.loc[[3, 5]].sum()

numeric_elements = pd.to_numeric(s, errors='coerce').dropna()
int_elem = numeric_elements[numeric_elements.astype(int) == numeric_elements]
variance = round(int_elem.var(), 2)

print("Новый Series с индексами от 2 до 11:")
print(new_s)
print("\nДисперсия целочисленных элементов:", variance)

Новый Series с индексами от 2 до 11:
2         1
3         2
4       3.1
5       hi!
6         5
7      -512
8     12.42
9      sber
10     10.1
11       98
dtype: object

Дисперсия целочисленных элементов: 59702.7


Исходя из кода выше, можно сказать, что из Series 3 и 5 просуммировать нельзя, т.к выдает ошибку из за нахождения строки на данном отрезке. А дисперсия в большей части приблизительна к ответу 4, поэтому **ответ: 3, 4**

#### Task 4.
1. Сгенерируйте Series из 100 значений нормально распределённой СВ (np.random.normal с дефолтными параметрами - нулевым средним и единичной дисперсией).
2. Возведите каждое значение серии в 3 степень, а значения индекса увеличьте в 3 раза.
3. Ответьте на следующие вопросы через запятую (без пробелов) (\*)
    - 1) Выведите сумму элементов, строго меньших 2.6, имеющих нечётные значения индекса.
    - 2) Выведите количество значений серии меньше нуля.

#### Пояснения:
(\*) Если получились ответы 3, 4.32, то в форму необходимо внести их в виде "3,4.32". То есть вещественные числа необходимо разделять точками. Не забудьте про фиксированный seed (его менять не нужно)!
 - Определенное значение seed нужно, чтобы ответы у всех выполняющих
это задание были одинаковые и их можно было проверить (так как генерируются
одинаковые series).
 - Следует внимательнее использовать [ ] для выбора данных по нескольким условиям:
либо выбирать данные последовательно, либо сразу по нескольким условиям,
но через оператор &. Отличие оператора and от оператора &: and - выводит
последнее проверенное значение, & - выводит пересечение значений.
Пример: s[ _  &  _ ].sum()

In [5]:
np.random.seed(242)
### Type your code here
series = pd.Series(np.random.normal(size=100))
series = series ** 3

series.index = series.index * 3

sum = series[(series.index % 2 != 0) & (series < 2.6)].sum()

negativ_values = len(series[series < 0])

print("{},{}".format(sum, negativ_values))


-15.150748101821666,51


Ответ: 1) -15.15 <br>
       2) 51




### Информация для последующих заданий
- Для всех последующих заданий будем использовать обезличенные транзакционные банковские данные. Для этого считайте в переменные tr_mcc_codes, tr_types, transactions и gender_train из одноимённых таблиц из папки data. Для таблицы transactions используйте только первые n=1000000 строк. Обратите внимание на разделители внутри каждого из файлов - они могут различаться!

In [11]:
### Type your code here
tr_mcc_codes = pd.read_table('tr_mcc_codes.csv', sep = ';')
tr_types = pd.read_table('tr_types.csv', sep = ';')
transactions = pd.read_table('transactions.csv', nrows = 1000000, sep = ',')
gender_train = pd.read_table('gender_train.csv', sep = ',')

## Описание данных
### Таблица ```transactions.csv```
##### Описание
Таблица содержит историю транзакций клиентов банка за один год и три месяца.

##### Формат данных

```
customer_id,tr_datetime,mcc_code,tr_type,amount,term_id
111111,15 01:40:52,1111,1000,-5224,111111
111112,15 15:18:32,3333,2000,-100,11122233
...
```
##### Описание полей

 - ```customer_id``` — идентификатор клиента;
 - ```tr_datetime``` — день и время совершения транзакции (дни нумеруются с начала данных);
 - ```mcc_code``` — mcc-код транзакции;
 - ```tr_type``` — тип транзакции;
 - ```amount``` — сумма транзакции в условных единицах со знаком; ```+``` — начисление средств клиенту (приходная транзакция), ```-``` — списание средств (расходная транзакция);
 - ```term_id``` — идентификатор терминала;


### Таблица ```gender_train.csv```

##### Описание
Данная таблица содержит информацию по полу для части клиентов, для которых он известен. Для остальных клиентов пол неизвестен.

##### Формат данных
```
customer_id,gender
111111,0
111112,1
...
```

##### Описание полей
 - ```customer_id``` — идентификатор клиента;
 - ```gender``` — пол клиента;

### Таблица ```tr_mcc_codes.csv```

##### Описание
Данная таблица содержит описание mcc-кодов транзакций.

##### Формат данных
```
mcc_code;mcc_description
1000;словесное описание mcc-кода 1000
2000;словесное описание mcc-кода 2000
...
```

##### Описание полей
 - ```mcc_code``` – mcc-код транзакции;
 - ```mcc_description``` — описание mcc-кода транзакции.

### Таблица ```tr_types.csv```

##### Описание
Данная таблица содержит описание типов транзакций.

##### Формат данных
```
tr_type;tr_description
1000;словесное описание типа транзакции 1000
2000;словесное описание типа транзакции 2000
...
```

##### Описание полей
 - ```tr_type``` – тип транзакции;
 - ```tr_description``` — описание типа транзакции;

#### Task 5.
1. В tr_types выберите произвольные 100 строк с помощью метода sample (указав при этом random_seed равный 242)
2. В полученной на предыдущем этапе подвыборке найдите долю наблюдений (стобец tr_description), в которой содержится подстрока 'плата' (в любом регистре). (\*)

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Строки "ПлатА за аренду", "ПлатАза аренду", "ПЛАТА" удовлетворяют условию, так как будучи переведёнными в нижний регистр содержат подстроку "плата".

In [12]:
### Type your code here
sampled_data = tr_types.sample(n=100, random_state=242)


fraction_contains = (sampled_data['tr_description'].str.contains('плата', case=False)).mean()

# Выводим ответ, округленный до двух знаков после запятой
print("{:.2f}".format(fraction_contains))

0.26


Ответ: 0.26

#### Task 6.
1. Для поля tr_type датафрейма transactions посчитайте частоту встречаемости всех типов транзакций tr_type в transactions.
2. Из перечисленных вариантов выберите те, которые попали в топ-5 транзакций по частоте встречаемости.

#### Выберите все верные пункты:
    - 1) Выдача наличных в АТМ Сбербанк России
    - 2) Комиссия за обслуживание ссудного счета
    - 3) Списание по требованию
    - 4) Оплата услуги. Банкоматы СБ РФ
    - 5) Погашение кредита (в пределах одного филиала)
    - 6) Покупка. POS ТУ СБ РФ

In [27]:
trans_count = transactions['tr_type'].value_counts()

top_5_trans = trans_count.head(5)

top_5_merge = pd.merge(pd.DataFrame(top_5_trans), tr_types, left_index=True, right_on='tr_type')



top_5_merge[['tr_description']].head(5)

tr_description
98                               Покупка. POS ТУ СБ РФ
106              Выдача наличных в АТМ Сбербанк России
101                             Покупка. POS ТУ Россия
99                      Оплата услуги. Банкоматы СБ РФ
129  Списание с карты на карту по операции <перевод...

**Ответ : 1,4,6**

#### Task 7.
1. В датафрейме transactions задайте столбец  customer_id в качестве индекса.
2. Выделите клиента с максимальной суммой транзакции (то есть с максимальным приходом на карту). (\*)
3. Найдите у него наиболее часто встречающийся модуль суммы приходов/расходов. (\**)


#### Выберите все верные пункты:
    - 1) 1122957.89
    - 2) 15721.41
    - 3) 22459.16
    - 4) 13475494.63
    - 5) 107407.78
    - 6) 65019.26

##### Пояснения:
(\*) Если у клиента были транзакции [-10000, 10, 0, -10], то максимумом будет являться значение 10. <br>
(\**) Если у клиента были транзакции [-10000, 10, 0, -10], то наиболее встречающийся модуль суммы транзакций равен 10, и встретился он 2 раза.

In [36]:
### Type your code here
transactions = pd.read_table('transactions.csv', index_col='customer_id', nrows = 1000000, sep = ',')

max_transaction_customer = transactions['amount'].idxmax()
max_transaction = transactions['amount'].max()

trans_grouped2 = transactions.loc[max_transaction_customer, 'amount'].abs().value_counts()
abs_max = trans_grouped2.idxmax()

print("Id клиента с максимальной суммой транзакции: ",max_transaction_customer)
print("Максимальная транзакция: ", max_transaction)
print("Средняя значение: ", abs_max)

Id клиента с максимальной суммой транзакции:  26782922
Максимальная транзакция:  145984525.17
Средняя значение:  22459.16


#### Task 8.
1. Найдите максимальную разницу между медианами суммы транзакций, посчитанными при заданных ниже условиях по полю amount из таблицы transactions (\*):
    - Медиана суммы транзакций
    - Медиана суммы транзакций по тем строкам, которые ни в одном из своих столбцов не содержат пустые значения
    - Медиана суммы транзакций по строкам, отсортированным по полю amount в порядке возрастания, и из которых удалены дублирующиеся по столбцам [mcc_code, tr_type] строки, причём при удалении соответстующих дублей остаются только последние из дублирующихся строк (keep='last')

Выведите ответ в виде вещественного числа, округлённого до двух знаков после запятой, отделив дробную часть точкой в формате "123.45"

##### Пояснения:
(\*) Для вычисления максимальной разницы между значениями списка можно использовать функцию np.ptp  
(\**) Если в результате получились значения [1,3,5], то макимальная разница между ними 4 == 5-1.

In [47]:
transactions['amount'] = pd.to_numeric(transactions['amount'], errors='coerce')

transactions = transactions.dropna(subset=['amount'])

med_sum = transactions['amount'].median()

transactions_noNan = transactions.dropna()
med_sum_withoutNan = transactions_noNan['amount'].median()

sorted_transactions = transactions.sort_values(by='amount', ascending=True)
drop_dup = sorted_transactions.drop_duplicates(subset=['mcc_code', 'tr_type'], keep='last')

med_sum_sort_nodup = drop_dup['amount'].median()

max_diff = np.max([abs(med_sum - med_sum_withoutNan), abs(med_sum - med_sum_sort_nodup), abs(med_sum_sort_nodup - med_sum_withoutNan)])


print(med_sum)
print(med_sum_withoutNan)
print(med_sum_sort_nodup)
print("{:.2f}".format(max_diff))

-5614.79
-5596.6
-920.83
4693.96
